### PR TITLE
fix: handle undefined configuration export

### PR DESCRIPTION
**Pull request by rishabh3112**
**What kind of change does this PR introduce?**
fix and test

**Did you add tests for your changes?**
Yup, did in TDD fashion.

**If relevant, did you update the documentation?**
Not required.

**Summary**
Treat `undefined` configuration as a case for `zero-config`

**Does this PR introduce a breaking change?**
Nope.

**Other information**
/cc @alexander-akait 
> when we have `webpack.config.js`  without `module.exports`  or `export`  we don't work, but no exports means zero configuration and should work

### CODE REVIEW

File: packages/webpack-cli/lib/webpack-cli.js [hunk 1]
```diff
@@ -67,10 +67,18 @@ class WebpackCLI {
...
-    if (result.default) {
+    if (result && result.default) {
       result = result.default;
     }
 
+    // Treat undefined configuration for zero-config
+    if (!result) {
+      this.logger.info(
+        "Configuration file provided didn't export any configuration, using defaults to build.",
+      );
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:78]: Let's avoid it, too aggressive logging, empty nothing = empty configuration, no need to logging about it

File: test/build/config/undefined/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1 @@
+module.exports = undefined;
```
Comment (alexander-akait) [inline test/build/config/undefined/webpack.config.js:1]: For test let's create just empty file, but keep this test too

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (rishabh3112) [reply packages/webpack-cli/lib/webpack-cli.js:78]: Okay, will remove.

File: test/build/config/undefined/webpack.config.js (hunk 2 above)
Comment (rishabh3112) [reply test/build/config/undefined/webpack.config.js:1]: Okay, will add a new test case.

File: packages/webpack-cli/lib/webpack-cli.js [hunk 3]
```diff
@@ -67,10 +67,15 @@ class WebpackCLI {
...
+    if (result && result.default) {
       result = result.default;
     }
 
+    // Treat undefined configuration for zero-config
+    if (!result) {
+      result = {};
+    }
+
     return result;
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:79]: Can be simplified by `return result || {}`;

Comment (alexander-akait) [reply packages/webpack-cli/lib/webpack-cli.js:79]: And lets add case where `module.exports.default = undefined`

Comment (rishabh3112) [reply packages/webpack-cli/lib/webpack-cli.js]: Okay, will do.

**Review by alexander-akait — DISMISSED** [review]

**Comment by rishabh3112** [comment]
@webpack/cli-team need approvals.

File: test/build/config/undefined-default/undefined-default.test.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+"use strict";
+const { resolve } = require("path");
+const { run } = require("../../../utils/test-utils");
+
+describe("config flag with undefined default export config file", () => {
+  it("should throw error with no configuration or index file", async () => {
+    const { exitCode, stderr, stdout } = await run(__dirname, [
+      "-c",
```
Comment (snitin315) [inline test/build/config/undefined-default/undefined-default.test.js:8]:
```suggestion
  it("should not throw error with no configuration or index file", async () => {
    const { exitCode, stderr, stdout } = await run(__dirname, [
      "--config",
```

File: test/build/config/no-code/no-code.test.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+"use strict";
+const { resolve } = require("path");
+const { run } = require("../../../utils/test-utils");
+
+describe("config flag with no code", () => {
+  it("should throw error with no configuration or index file", async () => {
+    const { exitCode, stderr, stdout } = await run(__dirname, [
+      "-c",
```
Comment (snitin315) [inline test/build/config/no-code/no-code.test.js:8]:
```suggestion
  it("should not throw error with no configuration or index file", async () => {
    const { exitCode, stderr, stdout } = await run(__dirname, [
      "--config",
```

File: test/build/config/undefined/src/index.js [hunk 6]
```diff
@@ -0,0 +1 @@
+console.log("Percy Weasley")
```
Comment (snitin315) [inline test/build/config/undefined/src/index.js:1]: newline at the end.

```suggestion
console.log("Percy Weasley");

```

File: test/build/config/undefined/undefined.test.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+"use strict";
+const { resolve } = require("path");
+const { run } = require("../../../utils/test-utils");
+
+describe("config flag with undefined export config file", () => {
+  it("should throw error with no configuration or index file", async () => {
+    const { exitCode, stderr, stdout } = await run(__dirname, [
+      "-c",
```
Comment (snitin315) [inline test/build/config/undefined/undefined.test.js:8]:
```suggestion
  it("should not throw error with no configuration or index file", async () => {
    const { exitCode, stderr, stdout } = await run(__dirname, [
      "--config",
```

**Review by snitin315 — COMMENTED** [review]
Few comments.

File: test/build/config/undefined/undefined.test.js (hunk 7 above)
Comment (rishabh3112) [reply test/build/config/undefined/undefined.test.js]: All tests seems to be using alias as well as not grammatically correct, will raise new PR to correct them in bulk.

File: test/build/config/undefined/src/index.js (hunk 6 above)
Comment (rishabh3112) [reply test/build/config/undefined/src/index.js]: Done

**Review by snitin315 — APPROVED** [review]

File: packages/webpack-cli/lib/webpack-cli.js [hunk 8]
```diff
@@ -67,11 +67,11 @@ class WebpackCLI {
     }
 
     // For babel/typescript
-    if (result.default) {
-      result = result.default;
+    if (result && typeof result === "object" && "default" in result) {
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:70]: Why we need `typeof result === "object"`?

Comment (rishabh3112) [reply packages/webpack-cli/lib/webpack-cli.js:70]: `result` can be of any type, thus we need to confirm it is object here to confirm we are handling babel/ts case and for usage of `in` operator we have to make sure `result` is `object`.

**Pull request merged**